File: apps/pass-at-k-study/benchmark/tasks/iris/l4_distributed_solver_step.cpp
// Task: l4_distributed_solver_step — matches L4DistributedSolverStep.
#include <cmath>
#include <cstdint>
#include <iostream>
#include <vector>

static std::uint32_t g_seed = 21u;
static float randf01() {
  g_seed = g_seed * 1103515245u + 12345u;
  return float(g_seed & 0xffffffu) / float(0x1000000u);
}

static float get(const std::vector<float> &A, int N, int i, int j) {
  i = (i % N + N) % N;
  j = (j % N + N) % N;
  return A[size_t(i) * N + j];
}

static bool run(int N) {
  std::vector<float> A(size_t(N) * N), B(size_t(N) * N);
  for (int i = 0; i < N; ++i)
    for (int j = 0; j < N; ++j) A[size_t(i) * N + j] = randf01();
  for (int i = 0; i < N; ++i)
    for (int j = 0; j < N; ++j) {
      float v = get(A, N, i, j);
      v += get(A, N, i - 1, j);
      v += get(A, N, i + 1, j);
      v += get(A, N, i, j - 1);
      v += get(A, N, i, j + 1);
      B[size_t(i) * N + j] = v / 5.f;
    }
  for (int t = 0; t < 5; ++t) {
    int i = (t * 19) % N;
    int j = (t * 29) % N;
    if (!std::isfinite(B[size_t(i) * N + j])) return false;
  }
  return true;
}

int main() {
  g_seed = 21u;
  if (!run(64)) {
    std::cout << "FAIL\n";
    return 1;
  }
  g_seed = 22u;
  if (!run(256)) {
    std::cout << "FAIL\n";
    return 1;
  }
  std::cout << "PASS\n";
  return 0;
}

File: apps/pass-at-k-study/benchmark/tasks/legate/l4_distributed_solver_step.py
# Task: l4_distributed_solver_step — matches L4DistributedSolverStep.
import math
import sys

_s = 21


def randf01() -> float:
    global _s
    _s = (_s * 1103515245 + 12345) & 0x7FFFFFFF
    return float(_s & 0xFFFFFF) / float(0x1000000)


def get(a, n, i, j):
    return a[i % n][j % n]


def run(n):
    a = [[randf01() for _ in range(n)] for _ in range(n)]
    b = [[0.0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            v = get(a, n, i, j)
            v += get(a, n, i - 1, j)
            v += get(a, n, i + 1, j)
            v += get(a, n, i, j - 1)
            v += get(a, n, i, j + 1)
            b[i][j] = v / 5.0
    for t in range(5):
        i = (t * 19) % n
        j = (t * 29) % n
        if not math.isfinite(b[i][j]):
            return False
    return True


def main() -> int:
    global _s
    _s = 21
    if not run(64):
        print("FAIL")
        return 1
    _s = 22
    if not run(256):
        print("FAIL")
        return 1
    print("PASS")
    return 0


if __name__ == "__main__":
    sys.exit(main())
